### ISSUE TITLE
`File watching not possible, no file descriptor left: 975`

### ISSUE THREAD

**Issue by jinnovation** [jinnovation/kele.el#89]

**Comment by jinnovation** [comment]
🤔 🧐  

<img width="586" alt="image" src="https://user-images.githubusercontent.com/1932579/212803424-0663d90d-b009-496d-b22d-53445225f96c.png">

**Comment by jinnovation** [comment]
This issue appears to happen when there are just **too many damn files in the discovery cache**. Kele tries to set a file watch on every file in there, causing us to hit the limit.

Realistically, it is not necessary to put a file-watch on **every single sub-directory in the discovery cache** -- only the ones corresponding to contexts **actually** in the user's kubeconfig.

**Comment by jinnovation** [comment]
Workaround is to delete the "stale" sub-dirs in the discovery cache.

**Comment by jinnovation** [comment]
Tying discovery-cache syncing with kubeconfig syncing would introduce an unfortunate dependency between `kele--global-kubeconfig-cache` and `kele--global-discovery-cache` but such is life. 🙃

**Comment by jinnovation** [comment]
[This bug blocks users for whom the discovery cache is large.](https://hachyderm.io/@jorj@jawns.club/109818068446739633) Worth addressing.

**Comment by jinnovation** [comment]
Hit up [@jorj@jawns.club](https://hachyderm.io/@jorj@jawns.club/109818068446739633) on Mastodon when we manage to get this fixed.

**Comment by jinnovation** [comment]
> Tying discovery-cache syncing with kubeconfig syncing would introduce an unfortunate dependency between kele--global-kubeconfig-cache and kele--global-discovery-cache but such is life. 

Two approaches come to mind that would allow us to severely curtail the number of file descriptors needed for discovery cache watching, while also not requiring us to couple discovery cache syncing to kubeconfig syncing.

## Approach 1: Lazy-loading w/ timed deletion

- On initialization, only set recursive file-watchers for the discovery cache sub-dir corresponding to a **single** context (this can be the current context);
- On switching to another context:
  - Set recursive file-watchers for the discovery cache sub-dir corresponding to that new context;
  - For the old context, set a timer that after N seconds cancels the file-watchers corresponding to that old context;
  - If the new context has a timer set, cancel it.

This approach allows us to watch discovery sub-cache for multiple contexts at a time, but has the following downsides:
- Complex and error-prone;
- Still leaves us open to the ticket issue, e.g. if user cycles through like 20 contexts in quick succession.

## Approach 2: Only watch a single discovery sub-cache at a time

This approach is far "simpler" but also more limited. The limitation, however, might be fine.

Basically, instead of attempting to file-watch multiple contexts' discovery sub-caches, **only file-watch the sub-cache corresponding to the current context**.

- On switching to a new context:
  - Cancel all existing discovery file-watches;
  - Wipe the existing `kele--discovery-cache` contents

Then, for actually loading the sub-cache contents (and, by extension, setting the file-watchers), we can load the contents + set file-watchers immediately after switching to a given context (largely reusing the existing async logic).

## Decision

Will likely go with Approach 2 here.

**Comment by jinnovation** [comment]
Whatever design we end up with here, def worth documenting in Explanations, e.g. "why not just cache the whole discovery cache?"
